### PR TITLE
feat(config): scoped secrets using pgp/gpg

### LINKED ISSUE #10641: Support encrypting longer secret through https://app.renovatebot.com/encrypt
**What would you like Renovate to be able to do?**

Encrypt longer strings. Currently, https://app.renovatebot.com/encrypt only supports input up to 245 bytes.

**Did you already have any implementation ideas?**

It may be possible to encrypt longer secrets using GPG https://github.com/openpgpjs/openpgpjs.

If this is not too complex, I can help out implementing using `openpgpjs`.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds PGP encryption.

## Context:

This solves two problems in one:
- Scoped secrets to eliminate replay attacks
- PGP-based encryption to remove the existing line length limitations

Closes #10641

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/decrypt.ts
Comment (viceice) [inline lib/config/decrypt.ts:23]: If you self-encrypt you need to remove those pgp comments? Can't they are optional?

This should at lease be documented how to generate the secrets

**Review by viceice — COMMENTED** [review]
one question, otherwise LGTM

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:375]: The "do things on your keyboard/mouse" part is shown twice, is that intentional?

File: lib/config/decrypt.ts
Comment (rarkins) [reply lib/config/decrypt.ts:23]: I might refactor it to make these optional

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:395]:
```suggestion
Any encrypted secrets using GPG must have a mandatory organization/group scope, and optionally can be scoped for a single repository only.
```

Not sure if you mean PGP or GPG... Confusing naming!

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:407]:
```suggestion
If you are migrating from the legacy public key encryption approach to use GPG, then move your legacy private key from `privateKey` to `privateKeyOld` and then put your new GPG private key in `privateKey`.
Doing so will mean that Renovate will first attempt to decrypt using the GPG key but fall back to the legacy key and try that next.
```

Again, do we mean PGP or GPG here? Confusing... 😄

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:324]: The rich diff preview is failing, I'm not sure what the reason for that is...

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some initial thoughts.

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:324]: Works for me

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:375]: It's from the log but I think we can remove it as it's not important

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md:324]: It's working for me now as well. 😄 Weird...

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:404]:
```suggestion
Used as an alternative to `privateKey`, if you want the key to be read from disk instead.
```

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:408]:
```suggestion
Used as an alternative to `privateKeyOld`, if you want the key to be read from disk instead.
```

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:412]:
```suggestion
Override this object if you want to change the URLs that Renovate links to, e.g. if you have an internal forum for asking for help.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I think we can use `want` instead of `wish to`, to simplify a bit.

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:325]: PR title says: `pgp`
In docs we use `gpg`

Not sure which PR title is actually correct here...

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:355]: Is it a smart idea to make a Renovate bot key that doesn't expire? I thought usually its a good idea to rotate keys like this? But maybe it's really inconvenient if the bot suddenly stops working because the key expired on a set schedule?

You probably know better than me what is a good idea and what isn't, so I'll let you decide. 😉

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:400]: I'm missing an explanation on when it's safe to remove the `privateKeyOld` config option when doing the "key rotation".

Maybe we want a user to do something like this?

1. Create new GPG key.
1. Put old key in `privateKeyOld` config option.
1. Confirm new GPG key works (how would you even confirm this if the old key keeps hanging around, though...)
1. Maybe wait until the PRs which use `privateOldKey` have been merged???
1. Remove `privateKeyOld` config option.

I'll let you respond and decide if this needs to go in the text. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I think it looks good overall, some things we might want to fix/look at though.

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:355]: It's ok if they want the key to expire. Ref: https://community.progress.com/s/article/Can-a-PGP-encryption-key-be-used-after-the-expiration-date

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:400]: They can remove it once all the old encrypted values have been migrated, or they wish to no longer support the old key and let repos fail

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs now.

**Review by viceice — COMMENTED** [review]
Just some docs things, code LGTM

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:382]: Yes, good point

Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:382]: Actually, the description above already says "do not configure a passphrase"

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
Will merge tomorrow as I need to synchronize things somewhat between docs, encrypt page, and app version

**Pull request merged**
